sirepo/template/template_common.py

```diff
@@ -83,6 +85,48 @@ def __init__(self, **kwargs):
         )
 
 
+class LogParser(PKDict):
+    def __init__(self, run_dir, log_filename, **kwargs):
+        super().__init__(run_dir=run_dir, log_filename=log_filename, **kwargs)
+        self.pksetdefault(
+            error_patterns=_DEFAULT_ERROR_PATTERNS,
+            multiline_error_pattern=r"^Traceback .*?^\w*Error: (.*?)\n",
+            default_msg="An unknown error occurred",
+        )
+
+    def parse_log_file_for_errors(self):
+        res = ""
+        p = self.run_dir.join(self.log_filename)
+        if not p.exists():
+            return res
+        with pkio.open_text(p) as f:
+            res = self._parse_log(f, res)
+        if res:
+            return res
+        return self.default_msg
+
+    def parse_multiline(self):
+        e = None
+        f = self.run_dir.join(self.log_filename)
+        if f.exists():
+            m = re.search(
+                self.multiline_error_pattern,
+                pkio.read_text(f),
+                re.MULTILINE | re.DOTALL,
+            )
+            if m:
+                e = m.group(1)
+        return e
+
+    def _parse_log(self, file, result):
```

Comment:
Don't pass result into this function. I think it is always `""`. It is dangerous to update values passed into functions (hard to know what state you are messing with). So better to avoid when possible.